File: locust/bronze.py
"""
BRONZE — Baseline: 50 concurrent users for 60 seconds
══════════════════════════════════════════════════════
Run with web UI (recommended for first time):
    locust -f locust/bronze.py --host http://localhost
    → open http://localhost:8089
    → set Users = 50, Spawn Rate = 5, click Start

Run headless (for screenshots / CI):
    locust -f locust/bronze.py \
           --headless --users 50 --spawn-rate 5 \
           --run-time 60s --host http://localhost \
           --html reports/bronze_report.html

Record from the output:
    - p95 latency  (Response Time column)
    - Failure rate (Failures/s column)
"""

import random

from locust import HttpUser, constant, task


class CashierUser(HttpUser):
    """
    Simulates one cashier terminal:
      - browses the product catalog (most common)
      - checks stock levels
      - rings up a sale (least common, always hits the DB)
    """

    wait_time = constant(1)  # 1-second pause between actions, same as k6's sleep(1)
    host = "http://localhost"

    @task(2)
    def browse_products(self):
        """50 % of requests — product catalog (Redis-cacheable, 300 s TTL)"""
        with self.client.get(
            "/api/products", name="GET /api/products", catch_response=True
        ) as resp:
            if resp.status_code != 200:
                resp.failure(f"Expected 200, got {resp.status_code}")

    @task(1)
    def check_inventory(self):
        """25 % of requests — stock levels (Redis-cacheable, 30 s TTL)"""
        with self.client.get(
            "/api/inventory", name="GET /api/inventory", catch_response=True
        ) as resp:
            if resp.status_code != 200:
                resp.failure(f"Expected 200, got {resp.status_code}")

    @task(1)
    def create_sale(self):
        """25 % of requests — checkout transaction (always writes to Postgres)"""
        with self.client.post(
            "/api/sales",
            json={
                "cashier_id": random.randint(1, 5),
                "items": [{"product_id": random.randint(1, 50), "quantity": 1}],
            },
            name="POST /api/sales",
            catch_response=True,
        ) as resp:
            if resp.status_code != 201:
                resp.failure(f"Expected 201, got {resp.status_code}")

File: k6/bronze.js
/**
 * BRONZE -- Baseline: 50 concurrent users for 60 seconds
 *
 * Run:    k6 run k6/bronze.js
 * Record: p(95) from http_req_duration  +  error_rate
 */
import http from "k6/http";
import { check, sleep } from "k6";
import { Rate, Trend } from "k6/metrics";

const errorRate = new Rate("error_rate");
const productLatency = new Trend("product_latency_ms", true);
const saleLatency = new Trend("sale_latency_ms", true);

export const options = {
  vus: 50,
  duration: "60s",
  thresholds: {
    http_req_duration: ["p(95)<3000"],
    error_rate: ["rate<0.05"],
  },
};

const BASE = "http://localhost";

export default function () {
  const roll = Math.random();

  if (roll < 0.5) {
    // 50% -- Browse product catalog (cacheable)
    const start = Date.now();
    const res = http.get(`${BASE}/api/products`);
    productLatency.add(Date.now() - start);
    errorRate.add(!check(res, { "products 200": (r) => r.status === 200 }));
  } else if (roll < 0.75) {
    // 25% -- Check inventory levels
    const res = http.get(`${BASE}/api/inventory`);
    errorRate.add(!check(res, { "inventory 200": (r) => r.status === 200 }));
  } else {
    // 25% -- Create a sale (DB write)
    const start = Date.now();
    const res = http.post(
      `${BASE}/api/sales`,
      JSON.stringify({
        cashier_id: Math.ceil(Math.random() * 5),
        items: [{ product_id: Math.ceil(Math.random() * 50), quantity: 1 }],
      }),
      { headers: { "Content-Type": "application/json" } },
    );
    saleLatency.add(Date.now() - start);
    errorRate.add(!check(res, { "sale 201": (r) => r.status === 201 }));
  }

  sleep(1);
}
